Write out stubs for most backing subsystems (#1199)
paritytech/polkadot: roadmap/implementors-guide/guide.md

#### Protocol

Input: None

Output:
  - Validation requests to Validation subsystem
  - `CandidateBackingMessage::Second`
  - Peer set manager: report peers (collators who have misbehaved)

#### Functionality

Overarching network protocol + job for every relay-parent

#### Candidate Selection Job

- Aware of validator key and assignment
- One job for each relay-parent, which selects up to one collation for the Candidate Backing Subsystem

----

### Statement Distribution

#### Description

The Statement Distribution Subsystem is responsible for distributing statements about seconded candidates between validators.

#### Protocol

#### Functionality

Implemented as a gossip protocol. Neighbor packets are used to inform peers which chain heads we are interested in data for. Track equivocating validators and stop accepting information from them. Forward double-vote proofs to the double-vote reporting system. Establish a data-dependency order:
  - In order to receive a `Seconded` message we must be working on corresponding chain head

>   - In order to receive an `Invalid` or `Valid` message we must have received the corresponding `Seconded` message.

Author: First I'll nitpick a bit: The model separates peers from validators. Peer B doesn't issue an approving vote, Validator X does. Whether that vote arrives via Peer A or B is irrelevant, as it whether Validator X controls peer B.

So I will clarify your question statement: We are talking about two validators X and Y. X has issued a `Seconded` statement of a candidate, and Y has issued a corresponding `Valid` statement.

Yes, we drop the `Valid` vote if we haven't seen the `Seconded` message. But we don't blacklist it forever. If we get the `Seconded` message later, we will then accept the `Valid` message.

But the situation outlined in your example cannot happen. Peer B can have an indication that we have seen the `Seconded` message one of two ways:
  1. We have sent them the `Seconded` message.
  2. They have sent us the `Seconded` message.

If Peer B has not seen a `Seconded` message from us, they should send the `Seconded` message before sending the `Valid` message, with the understanding that the `Valid` message will then be kept.

If Peer B has seen a `Seconded` message from us, they can just send the `Valid` message with the same understanding.

Now apply the transitive property. It is not possible for Peer B to have the `Valid` message with a corresponding `Seconded` message.

- Determine our validator index `i`, the set of backed candidates pending availability in `r`, and which bit of the bitfield each corresponds to.
- [TODO: wait T time for availability distribution?]
- Start with an empty bitfield. For each bit in the bitfield, if there is a candidate pending availability, query the availability store for whether we have the availability chunk for our validator index.
- For all chunks we have, set the corresponding bit in the bitfield.
- Sign the bitfield and dispatch a `BitfieldDistribution::DistributeBitfield` message.

----

### Collation Generation

[TODO]

#### Description

#### Protocol

#### Functionality

#### Jobs, if any

----

### Collation Distribution

[TODO]

#### Description

#### Protocol

#### Functionality

#### Jobs, if any

----

### Availability Store

#### Description

This is a utility subsystem responsible for keeping available certain data and pruning that data.

The two data types:
  - Full PoV blocks of candidates we have validated
  - Availability chunks of candidates that were backed and noted available on-chain.

For each of these data we have pruning rules that determine how long we need to keep that data available.

> PoV hypothetically only need to be kept around until the block where the data was made fully available is finalized. However, disputes can revert finality, so we need to be a bit more conservative. We should keep the PoV until a block that finalized availability of it has been finalized for 1 day (TODO: arbitrary, but extracting `acceptance_period` is kind of hard here...).

Author: #1176  should have more details

> PoV hypothetically only need to be kept around until the block where the data was made fully available is finalized. However, disputes can revert finality, so we need to be a bit more conservative. We should keep the PoV until a block that finalized availability of it has been finalized for 1 day (TODO: arbitrary, but extracting `acceptance_period` is kind of hard here...).

Author: Although as seen in https://github.com/paritytech/substrate/issues/6224 there is still some contention as to whether to allow reverting finality or not.

On block finality events:
  - TODO: figure out how we get block finality events from overseer
  - Handle all pruning based on the newly-finalized block.

On `QueryPoV` message:
  - Return the PoV block, if any, for that candidate hash.

On `QueryChunk` message:
  - Determine if we have the chunk indicated by the parameters and return it via the response channel if so.

On `StoreChunk` message:
  - Store the chunk along with its inclusion proof under the candidate hash and validator index.

----

### Candidate Validation

#### Description

This subsystem is responsible for handling candidate validation requests. It is a simple request/response server.

#### Protocol

Input:
  - CandidateValidation::Validate(CandidateReceipt, validation_code, PoV, response_channel)

#### Functionality

Given a candidate, its validation code, and its PoV, determine whether the candidate is valid. There are a few different situations this code will be called in, and this will lead to variance in where the parameters originate. Determining the parameters is beyond the scope of this module.

----

> ### Block Authorship (Provisioning)
> 
> #### Description
> 
> This subsystem is not actually responsible for authoring blocks, but instead is responsible for providing data to an external block authorship service beyond the scope of the overseer.

Reviewer: A subsystem which is not responsible for block authoring should not be called "Block Authorship". I think this is the same subsystem that in other PRs we've been calling "Candidate Backing.

```suggestion
### Candidate Backing (Provisioning)

#### Description

This subsystem is responsible for providing data to an external block authorship service beyond the scope of the overseer.
```
